### PR TITLE
EDU-3388: add more of a narrative to the converters page

### PR DESCRIPTION
**Pull request by axfelix**
From edu-3388, a request to bring this information closer to parity with our courses

### CODE REVIEW

File: docs/develop/go/converters-and-encryption.mdx [hunk 1]
```diff
@@ -18,11 +18,12 @@ tags:
   - Temporal SDKs
 ---
 
-This page shows how to do the following:
+By default, Temporal payloads are stored unencrypted inside of its data store.
```
Comment (mjameswh) [inline docs/develop/go/converters-and-encryption.mdx:21]: I'm a bit uncomfortable with how this idea is introduced. Nothing strictly wrong from a technical perspective, but I think we can give a much more positive feel to the subject.

- Starting this page with "By default" gives a first impression that this is some kind of option you need to turn on or configure. Then the rest of the section clarifies the idea, but at that point, the reader is left with the feeling that encryption is a missing/incomplete feature in Temporal, whereas it is that way by-design…

- From the Temporal Server's perspective, payloads are ALWAYS stored "unencrypted" in the data store (i.e. the server itself never add an extra layer of encryption; the data is stored as it received it). The Server don't even know either your data is encrypted or not.

I'd suggest introducing the idea from the following angle instead:

- Temporal's security model is designed around client-side encryption of Payloads, i.e. a client may encrypt Payloads _before_ sending them to the server (and decrypt them _after_ receiving them from the server). This provides a very high degree of confidentiality because the Temporal Server itself has absolutely no knowledge of the actual data. It also gives implementers more power and more freedom regarding which client is able to read which data (control access on keys, algorithms, etc).

- A Temporal developer adds client-side encryption of Payloads by providing a Custom Payload Codec to its Client. Depending on business needs, a complete implementation of Payload Encryption may involve selection of appropriate encryption algorithms, management of encryption keys, or restrict a subset of their users from viewing payload output. (etc)

- The server itself never adds encryption over Payloads. Therefore, unless client-side encryption is implemented, Payload data will be persisted in non-encrypted form to the data store, and any Client that can make requests to a Temporal namespace (including the Temporal UI and CLI) will be able to read Payloads contained in Workflows. When working with sensitive data, one should ALWAYS implement Payload encryption.

File: docs/develop/go/converters-and-encryption.mdx [hunk 2]
```diff
@@ -133,6 +144,11 @@ The order in which your encoding Payload Converters are applied depend on the or
 You can set multiple encoding Payload Converters to run your conversions.
 When the Data Converter receives a value for conversion, it passes through each Payload Converter in sequence until the converter that handles the data type does the conversion.
 
+Payload Converters can be customized independently of a Payload Codec.
+Temporal's Converter architecture looks like this:
+
+![Billing page showing Credits tab](/img/converter-architecture.png)
```
Comment (mjameswh) [inline docs/develop/go/converters-and-encryption.mdx:150]: The replacement text on this image is completely unrelated.

File: docs/develop/go/converters-and-encryption.mdx [hunk 3]
```diff
@@ -118,12 +119,22 @@ c, err := client.Dial(client.Options{
...
-For reference, see the following samples:
+- Data **encoding** is performed by the client using the default converter provided by Temporal or your custom Data Converter when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
+- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [temporal workflow show](/cli/workflow#show) command or when browsing the Web UI to view output.
 
-- [Codec server](https://github.com/temporalio/samples-go/tree/main/codec-server)
-- [Encryption](https://github.com/temporalio/samples-go/tree/main/encryption)
+For reference, see the [Encryption](https://github.com/temporalio/samples-go/tree/main/encryption) sample.
 
-## Use custom payload conversion {#custom-payload-conversion}
+### Using a Codec Server
```
Comment (mjameswh) [inline docs/develop/go/converters-and-encryption.mdx:127]:
```suggestion
### Using a Remote Payload Codec Server
```

File: docs/develop/go/converters-and-encryption.mdx [hunk 4]
```diff
@@ -118,12 +119,22 @@ c, err := client.Dial(client.Options{
 //...
 ```
 
-For reference, see the following samples:
+- Data **encoding** is performed by the client using the default converter provided by Temporal or your custom Data Converter when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
```
Comment (mjameswh) [inline docs/develop/go/converters-and-encryption.mdx:122]: We need to be careful about the terms. We can easily get confused.

This is taken out of an explanation I gave on Slack a few days ago:

> "Data Converters" is the high-level idea. The key concept here is that there are three major components to Data Converters:
> 
> **Payload Converters**: they handle serialization and deserialization of application objects; i.e. convert from JS objects to bytes in a Buffer. In TS, the default is to use JSON.stringify() and JSON.parse() which is normally enough.
> 
> **Failure Converters**: they play a role kind of similar to **Payload Converters**, but specifcially for Error/Failures objects. Let’s forget about this one for now.
> 
> **Payload Codecs**: they transform the byte representation of a payload, in and out; i.e. they convert a Buffer into another Buffer.
> 
> Now, Payload Converters and Failure Converters operate inside the context of the Workflow sandbox.
> 

This section is about Payload Codecs, but here, we're mixing in ideas that belong to Payload Converters.

But Payload Codecs operate outside of the sandbox, on your main Node’s thread.


Encryption and compression should be implemented as Payload Codecs, not Payload Converters.

Comment (mjameswh) [reply docs/develop/go/converters-and-encryption.mdx:122]: Actually, I think it would be desirable to give a quick top-down overview of Data Converters and its three components — Payload Converters/Failure Converters/Payload Codecs — very early on this page, before getting into code. The Data Converters setup code may look horribly complex if you don't understand this simple concept first.

**Review by mjameswh — CHANGES_REQUESTED** [review]
Gave few involving suggestions.

File: docs/develop/go/converters-and-encryption.mdx (hunk 1 above)
Comment (axfelix) [reply docs/develop/go/converters-and-encryption.mdx:21]: I like this a lot and I'm using it nearly verbatim, thanks!

File: docs/develop/go/converters-and-encryption.mdx [hunk 5]
```diff
@@ -18,11 +18,17 @@ tags:
   - Temporal SDKs
 ---
 
-This page shows how to do the following:
+Temporal's security model is designed around client-side encryption of Payloads.
+A client may encrypt Payloads before sending them to the server, and decrypt them after receiving them from the server.
+This provides a high degree of confidentiality because the Temporal Server itself has absolutely no knowledge of the actual data.
+It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access on keys, algorithms, etc.
```
Comment (brianmacdonald-temporal) [inline docs/develop/go/converters-and-encryption.mdx:24]:
```suggestion
It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access with keys, algorithms, or other security measures.
```

File: docs/develop/go/converters-and-encryption.mdx [hunk 6]
```diff
@@ -18,11 +18,17 @@ tags:
...
 
-- [Use a custom Payload Codec](#custom-payload-codec)
-- [Use custom payload conversion](#custom-payload-conversion)
-- [Use a custom Payload Converter](#custom-payload-converter)
+A Temporal developer adds client-side encryption of Payloads by providing a Custom Payload Codec to its Client.
+Depending on business needs, a complete implementation of Payload Encryption may involve the selection of appropriate encryption algorithms, management of encryption keys, and/or restriction a subset of their users from viewing payload output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/go/converters-and-encryption.mdx:27]:
```suggestion
Depending on business needs, a complete implementation of Payload Encryption may involve selecting appropriate encryption algorithms, managing encryption keys, restricting a subset of their users from viewing payload output, or a combination of these.
```

File: docs/develop/go/converters-and-encryption.mdx [hunk 7]
```diff
@@ -118,12 +124,22 @@ c, err := client.Dial(client.Options{
 //...
 ```
 
-For reference, see the following samples:
+- Data **encoding** is performed by the client using the converters and codecs provided by Temporal or your custom implementation when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
+- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [temporal workflow show](/cli/workflow#show) command or when browsing the Web UI to view output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/go/converters-and-encryption.mdx:128]:
```suggestion
- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [Temporal Workflow `show`](/cli/workflow#show) command or when browsing the Web UI to view output.
```

File: docs/develop/java/converters-and-encryption.mdx [hunk 8]
```diff
@@ -18,10 +18,17 @@ tags:
   - Temporal SDKs
 ---
 
-This page shows how to do the following:
+Temporal's security model is designed around client-side encryption of Payloads.
+A client may encrypt Payloads before sending them to the server, and decrypt them after receiving them from the server.
+This provides a high degree of confidentiality because the Temporal Server itself has absolutely no knowledge of the actual data.
+It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access on keys, algorithms, etc.
```
Comment (brianmacdonald-temporal) [inline docs/develop/java/converters-and-encryption.mdx:24]:
```suggestion
It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access with keys, algorithms, or other security measures.
```

File: docs/develop/java/converters-and-encryption.mdx [hunk 9]
```diff
@@ -18,10 +18,17 @@ tags:
...
 
-- [Use a custom Payload Codec in Java](#custom-payload-codec)
-- [Do custom Payload conversion in Java](#custom-payload-conversion)
+A Temporal developer adds client-side encryption of Payloads by providing a Custom Payload Codec to its Client.
+Depending on business needs, a complete implementation of Payload Encryption may involve the selection of appropriate encryption algorithms, management of encryption keys, and/or restriction a subset of their users from viewing payload output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/java/converters-and-encryption.mdx:27]:
```suggestion
Depending on business needs, a complete implementation of Payload Encryption may involve selecting appropriate encryption algorithms, managing encryption keys, restricting a subset of their users from viewing payload output, or a combination of these.
```

File: docs/develop/java/converters-and-encryption.mdx [hunk 10]
```diff
@@ -151,12 +158,22 @@ WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
               .build());
 ```
 
-For example implementations, see the following samples:
+- Data **encoding** is performed by the client using the converters and codecs provided by Temporal or your custom implementation when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
+- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [temporal workflow show](/cli/workflow#show) command or when browsing the Web UI to view output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/java/converters-and-encryption.mdx:162]:
```suggestion
- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [Temporal Workflow `show`](/cli/workflow#show) command or when browsing the Web UI to view output.
```

File: docs/develop/java/converters-and-encryption.mdx [hunk 11]
```diff
@@ -151,12 +158,22 @@ WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
...
+For reference, see the [Encryption](https://github.com/temporalio/samples-java/tree/main/core/src/main/java/io/temporal/samples/encryptedpayloads) sample.
+
+### Using a Codec Server
 
-- [Encrypted Payloads](https://github.com/temporalio/samples-java/tree/main/core/src/main/java/io/temporal/samples/encryptedpayloads)
-- [Remote Data Encoder and Codec Server](https://github.com/temporalio/sdk-java/tree/master/temporal-remote-data-encoder)
+A Codec Server is an HTTP server that uses your custom Codec logic to decode your data remotely.
+The Codec Server is independent of the Temporal Cluster and decodes your encrypted payloads through predefined endpoints.
+You create, operate, and manage access to your Codec Server in your own environment.
+The `temporal `CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```
Comment (brianmacdonald-temporal) [inline docs/develop/java/converters-and-encryption.mdx:171]:
```suggestion
The Temporal CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```

File: docs/develop/go/converters-and-encryption.mdx [hunk 12]
```diff
@@ -118,12 +124,22 @@ c, err := client.Dial(client.Options{
...
+For reference, see the [Encryption](https://github.com/temporalio/samples-go/tree/main/encryption) sample.
+
+### Using a Codec Server
 
-- [Codec server](https://github.com/temporalio/samples-go/tree/main/codec-server)
-- [Encryption](https://github.com/temporalio/samples-go/tree/main/encryption)
+A Codec Server is an HTTP server that uses your custom Codec logic to decode your data remotely.
+The Codec Server is independent of the Temporal Cluster and decodes your encrypted payloads through predefined endpoints.
+You create, operate, and manage access to your Codec Server in your own environment.
+The `temporal` CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```
Comment (brianmacdonald-temporal) [inline docs/develop/go/converters-and-encryption.mdx:137]:
```suggestion
The Temporal CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```

File: docs/develop/python/converters-and-encryption.mdx [hunk 13]
```diff
@@ -21,10 +21,17 @@ tags:
   - Temporal SDKs
 ---
 
-This page shows the following:
+Temporal's security model is designed around client-side encryption of Payloads.
+A client may encrypt Payloads before sending them to the server, and decrypt them after receiving them from the server.
+This provides a high degree of confidentiality because the Temporal Server itself has absolutely no knowledge of the actual data.
+It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access on keys, algorithms, etc.
```
Comment (brianmacdonald-temporal) [inline docs/develop/python/converters-and-encryption.mdx:27]:
```suggestion
It also gives implementers more power and more freedom regarding which client is able to read which data -- they can control access with keys, algorithms, or other security measures.
```

File: docs/develop/python/converters-and-encryption.mdx [hunk 14]
```diff
@@ -21,10 +21,17 @@ tags:
...
 
-- [How to use a custom Payload Codec](#custom-payload-codec)
-- [How to use a custom Payload Converter](#custom-payload-converter)
+A Temporal developer adds client-side encryption of Payloads by providing a Custom Payload Codec to its Client.
+Depending on business needs, a complete implementation of Payload Encryption may involve the selection of appropriate encryption algorithms, management of encryption keys, and/or restriction a subset of their users from viewing payload output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/python/converters-and-encryption.mdx:30]:
```suggestion
Depending on business needs, a complete implementation of Payload Encryption may involve selecting appropriate encryption algorithms, managing encryption keys, restricting a subset of their users from viewing payload output, or a combination of these.
```

File: docs/develop/python/converters-and-encryption.mdx [hunk 15]
```diff
@@ -86,10 +93,18 @@ client = await Client.connect(
 )
 ```
 
-For reference, see the following samples:
+- Data **encoding** is performed by the client using the converters and codecs provided by Temporal or your custom implementation when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
+- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [temporal workflow show](/cli/workflow#show) command or when browsing the Web UI to view output.
```
Comment (brianmacdonald-temporal) [inline docs/develop/python/converters-and-encryption.mdx:97]:
```suggestion
- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [Temporal Workflow `show`](/cli/workflow#show) command or when browsing the Web UI to view output.
```

File: docs/develop/python/converters-and-encryption.mdx [hunk 16]
```diff
@@ -86,10 +93,18 @@ client = await Client.connect(
...
+For reference, see the [Encryption](https://github.com/temporalio/samples-python/tree/main/encryption) sample.
+
+### Using a Codec Server
 
-- [Custom converter](https://github.com/temporalio/samples-python/tree/main/custom_converter)
-- [Encryption](https://github.com/temporalio/samples-python/tree/main/encryption)
+A Codec Server is an HTTP server that uses your custom Codec logic to decode your data remotely.
+The Codec Server is independent of the Temporal Cluster and decodes your encrypted payloads through predefined endpoints.
+You create, operate, and manage access to your Codec Server in your own environment.
+The `temporal `CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```
Comment (brianmacdonald-temporal) [inline docs/develop/python/converters-and-encryption.mdx:106]:
```suggestion
The Temporal CLI and the Web UI in turn provide built-in hooks to call the Codec Server to decode encrypted payloads on demand.
```

**Review by brianmacdonald-temporal — APPROVED** [review]
Additions look good to me. I suggested a few changes, mostly small, but repeated in each SDK.

File: docs/develop/dotnet/converters-and-encryption.mdx [hunk 17]
```diff
@@ -79,6 +87,19 @@ var myClient = await TemporalClient.ConnectAsync(new("localhost:7233")
 });
 ```
 
+- Data **encoding** is performed by the client using the converters and codecs provided by Temporal or your custom implementation when passing input to the Temporal Cluster. For example, plain text input is usually serialized into a JSON object, and can then be compressed or encrypted.
+- Data **decoding** may be performed by your application logic during your Workflows or Activities as necessary, but decoded Workflow results are never persisted back to the Temporal Cluster. Instead, they are stored encoded on the Cluster, and you need to provide an additional parameter when using the [temporal workflow show](/cli/workflow#show) command or when browsing the Web UI to view output.
```
[the remark this answers is not in the document]
Comment (axfelix) [reply docs/develop/dotnet/converters-and-encryption.mdx:91]: This one I'm keeping as is because it's more representative of the actual syntax -- other changes look good.
